File: src/main/java/com/freya02/botcommands/api/parameters/QuotableRegexParameterResolver.java
package com.freya02.botcommands.api.parameters;

import org.jetbrains.annotations.NotNull;

import java.util.regex.Pattern;

/**
 * Interface which indicates this class can resolve parameters for regex commands.
 * <br><b>Must be used with {@link RegexParameterResolver}.</b>
 */
public interface QuotableRegexParameterResolver<T extends ParameterResolver<T, R> & QuotableRegexParameterResolver<T, R>, R> extends RegexParameterResolver<T, R> {
	/**
	 * Returns a quoted pattern of the parameter resolver
	 *
	 * @return A quoted pattern of the original {@link RegexParameterResolver regex parameter resolver}
	 *
	 * @see RegexParameterResolver#getPattern()
	 */
	@NotNull
	Pattern getQuotedPattern();

	@NotNull
	@Override
	default Pattern getPreferredPattern() {
		return getQuotedPattern();
	}
}

File: src/main/kotlin/io/github/freya022/botcommands/api/parameters/QuotableRegexParameterResolver.kt
package io.github.freya022.botcommands.api.parameters

import io.github.freya022.botcommands.api.commands.prefixed.annotations.JDATextCommand
import java.util.regex.Pattern

/**
 * Parameter resolver for parameters of [@JDATextCommand][JDATextCommand].
 *
 * This resolver is an extension of [RegexParameterResolver], but the regex pattern is quoted to help command parsing.
 *
 * Needs to be implemented alongside a [ParameterResolver] subclass.
 *
 * @param T Type of the implementation
 * @param R Type of the returned resolved objects
 */
interface QuotableRegexParameterResolver<T, R : Any> : RegexParameterResolver<T, R>
        where T : ParameterResolver<T, R>,
              T : QuotableRegexParameterResolver<T, R> {
    /**
     * A quoted pattern of the parameter resolver.
     *
     * @see RegexParameterResolver.pattern
     */
    val quotedPattern: Pattern

    override val preferredPattern: Pattern
        get() = quotedPattern
}
